React tests

stevenphaedonos opened the pull request:
## Changes
I've implemented unit tests for the components/hooks implemented thus far in the React package.

A few notes:
- I have mocked the various posthog-js Feature Flag methods that the React package is dependent on.
- I had to add `@types/jest` as a dev dependency in the root `package.json`, otherwise the linter would throw type errors on commit.
- During the `prepublishOnly` script, I am running `yarn build` *before* `yarn test-react` as the React package is dependent on the built posthog-js module. Perhaps the following makes more sense?
```
"build": "yarn build-array && yarn build-module && yarn build-es-module && yarn process-types",
"prepublishOnly": "yarn lint && yarn test && yarn build && yarn test-react && yarn build-react",
```

## Checklist
- [x] Tests for new code (if applicable)
- [x] TypeScript definitions (module.d.ts) updated and in sync with library exports (if applicable)

macobo commented on react/src/hooks/__tests__/useFeatureFlags.test.tsx:
```diff
@@ -0,0 +1,70 @@
+        })
+    })
+
+    it('should refresh feature flags on an interval if a non-zero refreshInterval is provided', () => {
```
This is missing a test for no refreshInterval

stevenphaedonos replied:
Thanks, I've added the missing test case.

macobo commented on react/src/hooks/__tests__/useFeatureFlags.test.tsx:
```diff
@@ -0,0 +1,70 @@
+        })
+    })
+
+    it('should return `active` and `enabled` features when feature flags are changed', async () => {
```
How to rewrite this test using given2/no mocking:

```js

given('subject', () => renderHook(() => useFeatureFlags(), { wrapper: given.renderProvider }).result)

given('renderProvider', () => ({ children }) => <PostHogProvider client={given.posthog}>{children}</PostHogProvider>)

it('should return `active` and `enabled` features when feature flags are changed', () => {
  given('posthog', () => ({
    onFeatureFlags: (callback) => callback(ACTIVE_FEATURE_FLAGS),
    isFeatureEnabled: (flag) => flag !== 'example_feature_3'
  })
  expect(given.subject.current).toEqual({
       active: ACTIVE_FEATURE_FLAGS,
       enabled: ENABLED_FEATURE_FLAGS,
  })
})
```

Since you can override given at `it` or `describe` level this should result in much less boilerplate overall.

There's ways to cut the repetition down further if given2 is applied to the whole file (e.g. making onFeatureFlags behavior another given).

stevenphaedonos replied:
That's great, thank you. This is a much more straight-forward way to mock functions, especially given that they can be overriden. I've followed your advice and set the default mocks at the **describe** level, and override within the **test** only as necessary.

macobo reviewed (commented):
Thanks for the contribution.

Couple of things:
1. Let's move the jest-dependent dependencies into `react/package.json` and change test-react to `cd react; jest` instead. This avoids polluting the main package with unrelated code.
2. The tests are using a different style than main posthog-js tests, relying heavily on mocking. From past experience these sorts of tests are hard to work with and extend. I'd like to:
    - Introduce given2 to react tests
    - Drop types/jest, use jsx for tests over typescript. Typing in tests makes duck-typing really hard.

2 is slightly tricky - will post a few examples of this below. Also happy to do a pairing session for this refactoring or handle it myself.

stevenphaedonos commented:
> Thanks for the contribution.
> 
> Couple of things:
> 
>     1. Let's move the jest-dependent dependencies into `react/package.json` and change test-react to `cd react; jest` instead. This avoids polluting the main package with unrelated code.
> 
>     2. The tests are using a different style than main posthog-js tests, relying heavily on mocking. From past experience these sorts of tests are hard to work with and extend. I'd like to:
>        
>        * Introduce given2 to react tests
>        * Drop types/jest, use jsx for tests over typescript. Typing in tests makes duck-typing really hard.
> 
> 
> 2 is slightly tricky - will post a few examples of this below. Also happy to do a pairing session for this refactoring or handle it myself.

I agree, I was fighting against typescript in order to write the initial tests. It is much more straight-forward now after changing to jsx. 

I've updated the root package.json to be completely free of React-related jest dependencies, and updated the test-react script. The `prepublishOnly` script is still building both the original module and the React package at the same time. Does it make sense to make the following changes to `build` and `prepublishOnly` so that the React package is only built after it is tested?
```
"build": "yarn build-array && yarn build-module && yarn build-es-module && yarn process-types",
"prepublishOnly": "yarn lint && yarn test && yarn build && yarn test-react && yarn build-react",
```

Happy to resolve any further comments, and also happy to jump into a pairing session!

macobo commented:
Hello - I think this is basically ready to go. Merging this + creating a follow-up PR with a few minor tweaks.

Thanks again for the contribution! If you haven't reached out already, cc @yakkomajuri - let's send some merch your way.

Also @yakkomajuri - I think this needs docs up at posthog.com, how shall we handle that?
